Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -32,87 +34,171 @@ protocol SCCoAPTransportLayerDelegate: class {
-    fileprivate var udpSocketTag: Int = 0
-    
-    convenience init(port: UInt16) {
-        self.init()
-        self.port = port
+public final class SCCoAPUDPTransportLayer: NSObject {
+    weak public var transportLayerDelegate: SCCoAPTransportLayerDelegate!
+    var connections: [HostPortKey: NWConnection] = [:]
+    var listener: NWListener?
+
+    private func setupStateUpdateHandler(for connection: NWConnection, withHostPort hostPort: HostPortKey) -> NWConnection {
+        connection.stateUpdateHandler = { [weak self] newState in
+            switch newState {
+            case .failed(let error):
+                os_log("Connection FAILED", log: .default, type: .error, "\(error)")
+                self?.connections[hostPort]?.cancel()
+                self?.connections.removeValue(forKey: hostPort)
+                guard let self = self else { return }
+                self.transportLayerDelegate?.transportLayerObject(self, didFailWithError: error as NSError)
+            case .setup:
+                os_log("Connection entered SETUP state", log: .default, type: .info)
+            case .waiting(let reason):
+                os_log("Connection entered WAITING state. Reason %@", log: .default, type: .info, reason.debugDescription)
+            case .preparing:
+                os_log("Connection entered PREPAIRING state", log: .default, type: .info)
+            case .ready:
+                os_log("Connection entered READY state", log: .default, type: .info)
+                guard let self = self else { return }
+                self.startReads(from: connection, withHostPort: hostPort)
+            case .cancelled:
+                os_log("Connection entered is CANCELLED", log: .default, type: .info)
+                self?.connections.removeValue(forKey: hostPort)
+            @unknown default:
+                os_log("Connection is in UNKNOWN state", log: .default, type: .info)
+            }
+        }
+        return connection
     }
-    
-    fileprivate func setUpUdpSocket() -> Bool {
-        udpSocket = GCDAsyncUdpSocket(delegate: self, delegateQueue: DispatchQueue.main)
-        do {
-            try udpSocket!.bind(toPort: port)
-            try udpSocket!.beginReceiving()
-        } catch {
-            return false
+
+    private func mustGetConnection(forHost host: String, port: UInt16) -> NWConnection {
+        os_log("Currently %d NWConnection object(s) are alive", connections.count)
+        os_log("Getting connection object for HOST %@, PORT %d", log: .default, type: .info, host, port)
+        let connectionKey = HostPortKey(host: host, port: port)
+        if let connection = connections[connectionKey], connection.state != .cancelled {
+            os_log("Reusing existing NWConnection for HOST %@, PORT %d", log: .default, type: .info, host, port)
+            return connection
+        }
+        let connection = NWConnection(host: NWEndpoint.Host(host), port: NWEndpoint.Port(rawValue: port)!, using: .udp)
```

Comment:
`NWEndpoint.Port(rawValue: port)!`

Maybe we can do a better job than crashing on an invalid port, especially since we only know it at runtime.